### PR TITLE
Media Player and Spotify documentation improvements

### PR DESCRIPTION
**Pull request by kcd83**
## Proposed change

Documentation of existing functionality for `media_player` and `spotify` integrations. 

Describe the quirks of `source` from the Spotify API (name and forgetfulness) that can confuse users. Provider full example of various commands. Cross link relevant documentation (Spotify, Media Player, Scripts) to further help find relevant details.

## Type of change

- [x] Spelling, grammar or other readability improvements (`current` branch).
- [x] Adjusted missing or incorrect information in the current documentation (`current` branch).
- [ ] Added documentation for a new integration I'm adding to Home Assistant (`next` branch).
  - [ ] I've opened up a PR to add logo's and icons in [Brands repository](https://github.com/home-assistant/brands).
- [ ] Added documentation for a new feature I'm adding to Home Assistant (`next` branch).
- [ ] Removed stale or deprecated documentation.

## Additional information

- Link forum post: https://community.home-assistant.io/t/minor-improvements-to-the-spotify-integration-documentation/396168

## Checklist

- [x] This PR uses the correct branch, based on one of the following:
  - I made a change to the existing documentation and used the `current` branch.
  - I made a change that is related to an upcoming version of Home Assistant and used the `next` branch.
- [x] The documentation follows the Home Assistant documentation [standards][].

[standards]: https://developers.home-assistant.io/docs/documenting/standards

### CODE REVIEW

**Comment by homeassistant** [comment]
Hi @kcd83, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant.io#22003).

Once you do that we will be able to review and accept this pull request.

Thanks!

File: source/_integrations/media_player.markdown [hunk 1]
```diff
@@ -178,3 +178,64 @@ The way media players are displayed in the frontend can be modified in the [cust
 - `tv`: Device is a television type device.
 - `speaker`: Device is speaker or stereo type device.
 - `receiver`: Device is audio video receiver type device taking audio and outputting to speakers and video to some display.
+
+## Example script
+
+The example script below combines a number of media player commands to set the output speaker for [Spotify](https://www.home-assistant.io/integrations/spotify) to 30% volume and plays a randomly shuffled track from the playlist Calming Classical.
```
Comment (frenck) [inline source/_integrations/media_player.markdown]:
```suggestion
The example script below combines a number of media player commands to set the output speaker for [Spotify](/integrations/spotify) to 30% volume and plays a randomly shuffled track from the playlist Calming Classical.
```

File: source/_integrations/media_player.markdown [hunk 2]
```diff
@@ -178,3 +178,64 @@ The way media players are displayed in the frontend can be modified in the [cust
...
+
+This demonstrates how steps a user would go through manually can be combined into a script that can then be called by a single button press or automation.
+
+```yaml
+# configuration.yaml
+script:
```
Comment (frenck) [inline source/_integrations/media_player.markdown:193]: I am a fan of examples, but this is an example of a script. I'm not sure if that is the correct example (we rarely use scripts actually).

I think this makes more sense / be more clear, if shown as just a sequence or automation.

File: source/_integrations/media_player.markdown [hunk 3]
```diff
@@ -178,3 +178,64 @@ The way media players are displayed in the frontend can be modified in the [cust
...
+  demo_spotify:
```
Comment (frenck) [inline source/_integrations/media_player.markdown:194]: Please make sure this example complies with our YAML documentation style guide.

https://developers.home-assistant.io/docs/documenting/yaml-style-guide

File: source/_integrations/media_player.markdown [hunk 4]
```diff
@@ -178,3 +178,64 @@ The way media players are displayed in the frontend can be modified in the [cust
...
+      - service: media_player.media_next_track
+        data:
+          entity_id: media_player.spotify
+```
+
+Documentation:
+
+- [Scripts](https://www.home-assistant.io/integrations/script) configuration details
+- [Scripts Editor](https://www.home-assistant.io/docs/scripts/editor/) for a UI that reads and writes `scripts.yaml`
+- [Spotify](https://www.home-assistant.io/integrations/spotify) integration details
```
Comment (frenck) [inline source/_integrations/media_player.markdown]: Please do not use full links.

File: source/_integrations/spotify.markdown [hunk 5]
```diff
@@ -17,6 +17,8 @@ ha_platforms:
 
 The Spotify media player integration allows you to control [Spotify](https://www.spotify.com/) playback from Home Assistant.
 
+Once an output [source]](#selecting-output-source) is selected you will be able to play content from Spotify using the standard Media Browser UI or `media_player` service.
```
Comment (frenck) [inline source/_integrations/spotify.markdown]: I don't think this belongs in the introduction

File: source/_integrations/spotify.markdown [hunk 6]
```diff
@@ -113,6 +115,32 @@ script:
 
 The `media_content_id` value can be obtained from the Spotify desktop app by clicking on the more options ("...") next to the album art picture, selecting "Share" and then "Copy Spotify URI" or "Copy Playlist Link" (also available in the Spotify phone and web app).
 
+The `media_content_id` can be a URL (e.g. `https://open.spotify.com/playlist/37i9dQZF1DWVFeEut75IAL`) or the Spotify URI string (e.g. `spotify:playlist:37i9dQZF1DWVFeEut75IAL`). (play_media)[https://www.home-assistant.io/integrations/media_player/#service-media_playerplay_media] requires the correct corresponding `media_content_type`.
```
Comment (frenck) [inline source/_integrations/spotify.markdown]: This is already in the line above?

File: source/_integrations/spotify.markdown [hunk 7]
```diff
@@ -113,6 +115,32 @@ script:
...
+
+## Selecting output source
+
+To play media Spotify needs a device selected for audio output known as the `source`. Currently, this integration cannot initiate playback to a device not already available from the Spotify API. You need to first use Spotify on your smart speaker or phone to populate the source list.
```
Comment (frenck) [inline source/_integrations/spotify.markdown]: This makes it sound like this is an integration issue, which it is not. It is a limitation of the Spotify API.

> You need to first use Spotify on your smart speaker or phone to populate the source list.

That won't work as intuitively as described. It will take quite some time before that becomes visible. Instead, one should better be using the capabilities of the device and its specific integrations provided to use Spotify.

File: source/_integrations/spotify.markdown [hunk 8]
```diff
@@ -113,6 +115,32 @@ script:
...
+
+```yaml
+# Example code to select an AV receiver as the output device
+service: media_player.select_source
+entity_id: media_player.spotify
+data:
+  source: "Denon AVR-X2000"
+```
+
+The device name must be known to Spotify and therefore is often different from names given to entities in Home Assistant.
```
Comment (frenck) [inline source/_integrations/spotify.markdown]: What does this mean?

File: source/_integrations/spotify.markdown [hunk 9]
```diff
@@ -113,6 +115,32 @@ script:
...
+
+The source list of available devices can be found in the Details section of the Spotify Media Player Control and the `source_list` attribute in the Developer tools States or templates using:
+
+```jinja
+{{ state_attr('media_player.spotify', 'source_list') }}
+```
```
Comment (frenck) [inline source/_integrations/spotify.markdown]: This is absolutely not needed, as there are other, more use-friendly, options available.

File: source/_integrations/spotify.markdown [hunk 10]
```diff
@@ -113,6 +115,32 @@ script:
...
+# Example code to select an AV receiver as the output device
+service: media_player.select_source
+entity_id: media_player.spotify
+data:
+  source: "Denon AVR-X2000"
+```
+
+The device name must be known to Spotify and therefore is often different from names given to entities in Home Assistant.
+
+The source list of available devices can be found in the Details section of the Spotify Media Player Control and the `source_list` attribute in the Developer tools States or templates using:
```
Comment (frenck) [inline source/_integrations/spotify.markdown]: Let's link to some using My Home Assistant links.

File: source/_integrations/spotify.markdown [hunk 11]
```diff
@@ -113,6 +115,32 @@ script:
...
+
+```jinja
+{{ state_attr('media_player.spotify', 'source_list') }}
+```
+
+<div class='note'>
+  A known limitation with Spotify is that devices disappear from the source list over time. Many Spotify Connect devices tend to persist, other devices such as Amazon Alexa speakers may only be remembered for a while, and phones and Google Cast devices disappear as soon as playback stops.
```
Comment (frenck) [inline source/_integrations/spotify.markdown]: Let's change this to something more concrete instead of "magically disappearing".

e.g., 

> Spotify Connect devices are only visible in the list of media sources when they are actively used with the connected Spotify account.

**Review by frenck — CHANGES_REQUESTED** [review]
Hi @kcd83 👋 

Thanks for the additions! I've left some comments, could you take a look?

File: source/_integrations/spotify.markdown (hunk 7 above)
Comment (kcd83) [reply source/_integrations/spotify.markdown]: Your comment here makes me think I didn't write this clearly so yes I will amend it.

> That won't work as intuitively as described...

I open the Spotify app and select a device from there. Alternatively I tell Google Home to cast Spotify. Both ways make the device available intermediately. Are you suggesting some other way like HACS Spotcast or was my description not clear?


Good point about where the limitation comes from. Perhaps it should say "Currently, the Spotify API cannot initiate playback to a device not already known to the Spotify and in the source list"

**Comment by kcd83** [comment]
Thanks for the detailed comments. I will go through them ASAP

**Comment by homeassistant** [comment]
Hi @kcd83, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant.io#22003).

Once you do that we will be able to review and accept this pull request.

Thanks!

**Comment by homeassistant** [comment]
Hi @kcd83, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant.io#22003).

Once you do that we will be able to review and accept this pull request.

Thanks!

**Comment by homeassistant** [comment]
Hi @kcd83, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant.io#22003).

Once you do that we will be able to review and accept this pull request.

Thanks!

File: source/_integrations/spotify.markdown (hunk 8 above)
Comment (kcd83) [reply source/_integrations/spotify.markdown]: It means if Spotify has my phone named "Kevs S10 but Home Assistant calls it "SM-G973F", the latter is irrelevant.

I think I over emphasis that this is the name in the Spotify API, _not_ Home Assistant. Initially it was not obvious to me that 
- HA tells Spotify API to play on a Device
- NOT HA streams to a Device from Spotify API 

But all of that is covered below so I will simply this 😄

File: source/_integrations/spotify.markdown (hunk 11 above)
Comment (kcd83) [reply source/_integrations/spotify.markdown]: I like the positive wording you have here. I will drop the vague reference to particular brands as it is hard to verify.

Of those that do persist (not phones or Google Cast) it seems unclear why devices do "magically disappear". I think someone said Spotify only recalls the last 3 connect devices but I cannot find this documented anywhere

File: source/_integrations/spotify.markdown (hunk 10 above)
Comment (kcd83) [reply source/_integrations/spotify.markdown]: Good idea. I cannot how to do this in the [docs](https://my.home-assistant.io/create-link/) but I found and example to copy 🤞

**Comment by kcd83** [comment]
I do not understand the Netlify test failures. One error is

> Liquid syntax error (line 204): Expected end_of_string but found id in "{{ not is_state('media_player.spotify', 'idle') }}"

But this jinja template is correctly quoted. All other errors also point to this one log file so I cannot see how to debug them further.

File: source/_integrations/media_player.markdown (hunk 2 above)
Comment (kcd83) [reply source/_integrations/media_player.markdown:193]: Interesting, in my mind a script is just an automation without a trigger and I'm not sure what simple trigger would suit this example.

I like that the script makes it standalone and provides tracing/debugging tools.  Even though it is a little less useful in real usage (without also adding a button in lovelace) I find it easier to test with. What do you and others use?

I really like how you can run individual script actions in the editor and I have struggled to find somewhere else you can just copy and paste a sequence. Especially without editing yaml arrays/indentation which seems to continually confuse newbies.

Sorry, lots of words. I genuinely want this to be as useful to new users as possible 😃

File: source/_integrations/spotify.markdown (hunk 5 above)
Comment (kcd83) [reply source/_integrations/spotify.markdown]: You are right the source detail doesn't belong here. As a new user I didn't even realise where to look next, so how about links?

I will push what I mean

**Comment by kcd83** [comment]
Thank you for your time spent going though my PR @frenck. Your suggestions help.

There are a few things that require more feedback. Hopefully these don't take too much time to review

**Comment by kcd83** [comment]
TL;DR this type of contribution is not welcome. It went into too many quirks outside basic functionality for a quick review
